pkg/providers/gcp/cluster_network_provider.go

```diff
@@ -0,0 +1,545 @@
+	NetworkApi  gcpiface.NetworksAPI
+	SubnetApi   gcpiface.SubnetsApi
+	ServicesApi gcpiface.ServicesAPI
+	AddressApi  gcpiface.AddressAPI
+	Logger      *logrus.Entry
+	ProjectID   string
+}
+
+type CreateVpcInput struct {
+	CidrBlock string
+}
+
+// NewNetworkManager initialises all required clients
+func NewNetworkManager(ctx context.Context, projectID string, opt option.ClientOption, client client.Client, logger *logrus.Entry) (*NetworkProvider, error) {
+	networksApi, err := gcpiface.NewNetworksAPI(ctx, opt)
+	if err != nil {
+		return nil, errorUtil.Wrap(err, "Failed to initialise network client")
+	}
+	subnetsApi, err := gcpiface.NewSubnetsAPI(ctx, opt)
+	if err != nil {
+		return nil, errorUtil.Wrap(err, "Failed to initialise subnetworks client")
+	}
+	servicesApi, err := gcpiface.NewServicesAPI(ctx, opt)
+	if err != nil {
+		return nil, errorUtil.Wrap(err, "Failed to initialise servicenetworking client")
+	}
+	addressApi, err := gcpiface.NewAddressAPI(ctx, opt)
+	if err != nil {
+		return nil, errorUtil.Wrap(err, "Failed to initialise addresses client")
+	}
+	if logger == nil {
+		logger = logrus.NewEntry(logrus.StandardLogger())
+	}
+	return &NetworkProvider{
+		Client:      client,
+		NetworkApi:  networksApi,
+		SubnetApi:   subnetsApi,
+		ServicesApi: servicesApi,
+		AddressApi:  addressApi,
+		Logger:      logger.WithField("provider", "gcp_network_provider"),
+		ProjectID:   projectID,
+	}, nil
+}
+
+func (n *NetworkProvider) CreateNetworkIpRange(ctx context.Context, cidrRange *net.IPNet) (*computepb.Address, error) {
+	// build ip address range name
+	ipRangeName, err := resources.BuildInfraName(ctx, n.Client, defaultIpRangePostfix, defaultGcpIdentifierLength)
+	if err != nil {
+		return nil, errorUtil.Wrap(err, "failed to build ip address range infra name")
+	}
+	address, err := n.getAddressRange(ctx, ipRangeName)
+	if err != nil {
+		return nil, errorUtil.Wrap(err, "failed to retrieve ip address range")
+	}
+	// if it does not exist, create it
+	if address == nil {
+		clusterVpc, err := getClusterVpc(ctx, n.Client, n.NetworkApi, n.ProjectID, n.Logger)
+		if err != nil {
+			return nil, errorUtil.Wrap(err, "failed to get cluster vpc")
+		}
+		subnets, err := n.getClusterSubnets(ctx, clusterVpc)
+		if err != nil {
+			return nil, errorUtil.Wrap(err, "failed to get cluster subnetworks")
+		}
+		err = validateCidrBlock(cidrRange, subnets)
+		if err != nil {
+			return nil, errorUtil.Wrap(err, "ip range validation failure")
+		}
+		err = n.createAddressRange(ctx, clusterVpc, ipRangeName, cidrRange)
+		if err != nil {
+			return nil, errorUtil.Wrap(err, "failed to create ip address range")
+		}
+		if address, err = n.waitForAddress(ctx, ipRangeName); err != nil {
+			return nil, errorUtil.Wrap(err, "failed to retrieve ip address range")
+		}
+	}
+	return address, nil
+}
+
+func (n *NetworkProvider) waitForAddress(ctx context.Context, ipRangeName string) (*computepb.Address, error) {
+	addressChan := make(chan *computepb.Address, 1)
```

Comment:
I believe this does block the primary reconcile up to the defaultTimeout duration. I can't quite remember if there was a specific reason this was added, do you happen to recall @cecobask?

Reply:
I believe that the standard for operators is to return and requeue the reconcile rather than blocking it somehow. we do that in the case of the CredentialsRequest for example.